src/recipes/editor-plugin/inspector-plugins.md

```diff
@@ -0,0 +1,343 @@
+```
+
+Let's add a handler for the button:
+
+```rust
+#[godot_api]
+impl RandomNumberEditor {
+    #[func]
+    fn handle_press(&mut self) {
+        // Update value by button click
+        // - take prop name, randomize number
+        // - send prop name and random number to godot engine for update value
+        // - update button text
+        let property_name = self.base().get_edited_property();
+        let num = rand::thread_rng().gen_range(0..100);
+        godot_print!("Randomize! {} for {}", num, property_name);
+        self.base_mut()
+            .emit_changed(property_name, num.to_variant().to());
+        if let Some(mut button) = self.button.clone() {
+            let text = format!("Randomize: {}", num);
+            button.set_text(text.into());
+        } else {
+            // Print error of something went wrong
+            godot_error!("Button wasn't found in handle_press");
+        }
+    }
+}
+```
+
+
+## Add Inspector plugin
+
+Now we need to connect this editor to fields with an integer type.
+To do this, we need to create an EditorInspectorPlugin.
+
+```rust
+#[derive(GodotClass)]
+#[class(tool, init, base=EditorInspectorPlugin)]
+struct RandomInspectorPlugin {
+    base: Base<EditorInspectorPlugin>,
+}
+```
+
+IEditorInspectorPlugin implementation:
+
+```rust
+#[godot_api]
+impl IEditorInspectorPlugin for RandomInspectorPlugin {
+    fn parse_property(
+        &mut self,
+        _: Gd<classes::Object>, // inspect object
+        type_: VariantType, // value type
+        name: GString, // property name
+        _: global::PropertyHint, // hint type
+        _: GString, // hit name
+        _: global::PropertyUsageFlags, // property usage flag
+        _: bool, // wide
+    ) -> bool {
+        if type_ == VariantType::INT {
+            let editor = RandomNumberEditor::new_alloc().to_variant().to();
+            self.base_mut().add_property_editor(name, editor);
+            return true;
+        }
+
+        false
+    }
+
+    fn can_handle(&self, _: Gd<classes::Object>) -> bool {
+        true
+    }
+}
+```
+
+If `parse_property` returns `true`, the editor will be created and replace the current
+representation; if not, it's necessary to return `false`.
+This allows for specific control over where and how processing occurs.
+
+
+## Add Editor Plugin
+
+Only one thing left to do: define the editor plugin that will kick off all this magic!
+This can be a generic EditorPlugin or a more specific InspectorEditorPlugin, depending
+on what you want to achieve.
+
+
+```rust
+#[derive(GodotClass)]
+#[class(tool, init, editor_plugin, base=EditorPlugin)]
+struct RustEditorPlugin {
+    base: Base<EditorPlugin>,
+    random_inspector: Gd<RandomInspectorPlugin>,
+}
+```
+
+```rust
+#[godot_api]
+impl IEditorPlugin for RustEditorPlugin {
+    fn enter_tree(&mut self) {
+        // create our inspector plugin and save it to next remove
+        self.random_inspector = RandomInspectorPlugin::new_gd();
+        self.to_gd()
```

Comment:
```suggestion
        self.base_mut()
```

Reply:
then I'll get:
```
error[E0502]: cannot borrow `self.random_inspector` as immutable because it is also borrowed as mutable
  --> src\addon.rs:23:35
   |
22 |         self.base_mut()
   |         ---------------
   |         |
   |         mutable borrow occurs here
   |         a temporary with access to the mutable borrow is created here ...
23 |             .add_inspector_plugin(self.random_inspector.to_variant().to());
   |                                   ^^^^^^^^^^^^^^^^^^^^^                   - ... and the mutable borrow might be used here, when that temporary is dropped and runs the destructor for type `BaseMut<'_, RustEditorPlugin>`
   |                                   |
   |                                   immutable borrow occurs here

```

Reply:
You could create the argument in a separate variable, then it should work.

Reply:
yes, but what the difference? could you explain cons and pros?

Reply:
`base()/base_mut()` preserve type safety, and can often avoid runtime errors that `to_gd()` wont be able to catch. The simplest example is this:
```rs
fn ready(&mut self) {
  // this will error at runtime because `on_notification` tries to get another mutable reference to self
  self.to_gd().add_child(Node::new());
  // whereas this wont because `base_mut` avoids that problem
  self.base_mut().add_child(Node::new());
}

fn on_notification(&mut self, what: NodeNotification) {

}
```

Reply:
thank you